[opentitanlib] Add SPHINCS+ signing and verification

jon-flatley opened the pull request:
This PR adds the ability to sign and verify messages using the shake-128s-simple variant of SPHINCS+ to opentitanlib.

Leaving this as a draft until the following are addressed:

- [ ] Add tests, particularly to make sure the IO implementations interact correctly with our `big_num` utils.
- [ ] Decide if the way the `pqcrypto-sphincsplus` crate is treated by bazel is tolerable. See issues relating to this in a [comment by @cfrantz](https://github.com/lowRISC/opentitan/issues/17933#issuecomment-1504275973).
- [ ] Sanity check the implementation on an OT image (requires more work to inject the signature into the manifest.)

cfrantz commented on sw/host/opentitanlib/src/image/image.rs:
```diff
@@ -150,6 +143,16 @@ impl Image {
+    where
+        F: FnOnce(&[u8]) -> R
+    {
+        f(&self.data.bytes[offset_of!(Manifest, rsa_signature)..self.size])
```
I thought in recent discussions we decided that the signed region should _not_ include the `rsa_signature`.

jon-flatley replied:
I'm not sure, my understanding is that `rsa_signature` is included. @alphan can you weigh in on this?

alphan replied:
Yeah, we decided to exclude `rsa_signature` from the signed region. Apologies for the delay in updating you @jon-flatley.

cfrantz commented on sw/host/opentitanlib/src/crypto/spxplus.rs:
```diff
@@ -0,0 +1,164 @@
+
+impl ToWriter for SpxKeypair {
+    fn to_writer(&self, w: &mut impl Write) -> Result<()> {
+        // Write out the keypair as a fixed length byte-string consisting of the public key
+        // concatenated with the secret key.
+        w.write_all(self.pk.as_bytes())?;
+        w.write_all(self.sk.as_bytes())?;
```
Is the on-disk form of a Spx key or keypair just a bag of bytes? (ie: is there no file format?)

Should we consider a PEM container with a headers like "SPHINCS+ PUBLIC KEY" and the like?  The [pem_rfc7468 crate](https://crates.io/crates/pem-rfc7468) makes this sort of thing pretty easy:

```
// Turn your key into a bag of bytes:
let key = whatever_conversion_to_get_byte_buffer(); // (e.g. [u8; N] or Vec<u8>)
let pem = pem_rfc7468::encode_string("SPHINCS+ KEY PAIR", LineEnding::LF, &key)?;
w.write_all(pem.as_bytes())?;
```

Perhaps the decoding functionality is more useful, as you get back the "kind" so you can verify you've read the correct type of PEM file:

```
let mut data = Vec::new();
r.read_to_end(&mut data)?;
let (kind, key) = pem_rfc7468::decode_vec(&data)?;
if kind == "SPHINCS+ KEY PAIR" {
    Ok(SpxKeypair { ... })
} else {
    Err("The wrong kind of thing");
}
```

jon-flatley replied:
I agree that a PEM file is much nicer than the bag of bytes. It's my understanding that there is no format for SPHINCS+ keys, and that the format isn't all that important at the moment, so long as opentitantool is consistent. I'm in favor of this, it seems like a much better approach.

alphan replied:
There is no standard format yet. We floated binary blobs, hjson files (which can also encode type and other info) but PEM SGTM too. I'd be OK with addressing this in a follow-up PR.

cfrantz commented on sw/host/opentitantool/src/command/image.rs:
```diff
@@ -19,6 +19,7 @@ use opentitanlib::crypto::sha256::Sha256Digest;
 use opentitanlib::image::image::{self, ImageAssembler};
 use opentitanlib::image::manifest_def::ManifestSpec;
 use opentitanlib::util::parse_int::ParseInt;
+use opentitanlib::util::file::{FromReader, ToWriter};
```
Are we using this yet?

cfrantz reviewed (commented):
This looks pretty good to me.

If you can, could you reformat the stack of commits?  There seems to be a lot of intermediate work in the stack:
- Changing the crate.annotation between commits.
- Implementing `io::Read/Write` then replacing them with `FromReader/ToWriter`.

I think it would be fine to squash this PR down to two or three commits:
1) Bazel stuff to get pqcrypto added.
2a) Everything else, or:
2b) A commit with key handling (loading, saving) and a commit with sign/verify.

jon-flatley commented:
Thanks for the review @cfrantz, find my responses to you comments inline below.

> This looks pretty good to me.
> 
> If you can, could you reformat the stack of commits? There seems to be a lot of intermediate work in the stack:
> 
>     * Changing the crate.annotation between commits.

My mistake, I'll clean the commit history up.

> 
>     * Implementing `io::Read/Write` then replacing them with `FromReader/ToWriter`.

Is this what we want? In a previous version of this PR that's what I did, but it didn't feel right. What's really being done here is serialization/deserialization, so I don't know if structs implementing these traits can be thought of as writers/readers. I don't know what the idiomatic solution is here to be honest. Maybe using `serde`'s `serialize_bytes()`?
> 
> 
> I think it would be fine to squash this PR down to two or three commits:
> 
>     1. Bazel stuff to get pqcrypto added.
>        2a) Everything else, or:
>        2b) A commit with key handling (loading, saving) and a commit with sign/verify.

Sounds good to me!

cfrantz commented:
> Is this what we want? In a previous version of this PR that's what I did, but it didn't feel right. What's really being done here is serialization/deserialization, so I don't know if structs implementing these traits can be thought of as writers/readers. I don't know what the idiomatic solution is here to be honest. Maybe using `serde`'s `serialize_bytes()`?

I like FromReader/ToWriter - they make reading/writing files easy and they're adaptable to anything else that uses the `Read/Write` traits.  I don't think we need to involve `serde` here unless we want to make the externalized form json or something.  

I think using PEM for the externalized form makes sense, but I'm fretting a bit about what to use as the PEM identifier.  I believe the typical PEM encoding of an RSA key is a base64 representation of the DER encoding.  If there aren't any OIDs for spx keys, then there probably is no valid DER encoding.

I'm thinking about `RAW SPHINCS+ PUBLIC KEY` and `RAW SPHINCS+ PRIVATE KEY` or similar.  WDYT?

jon-flatley commented:
> I'm thinking about RAW SPHINCS+ PUBLIC KEY and RAW SPHINCS+ PRIVATE KEY or similar. WDYT?

I'm not sure what the best naming scheme is. Right now two files get generated, one with just the public key and the other with both the public and secret key. Are you envisioning the "PRIVATE KEY" containing the public key + secret key pair or just the secret key? In case of the latter, I think we should go with "SECRET KEY".

cfrantz commented:
> > I'm thinking about RAW SPHINCS+ PUBLIC KEY and RAW SPHINCS+ PRIVATE KEY or similar. WDYT?
> 
> I'm not sure what the best naming scheme is. Right now two files get generated, one with just the public key and the other with both the public and secret key. Are you envisioning the "PRIVATE KEY" containing the public key + secret key pair or just the secret key? In case of the latter, I think we should go with "SECRET KEY".

Yes - I think the "PRIVATE KEY" file should contain both, just as the SpxKeyPair struct does (and an "RSA PRIVATE KEY" contains both the private and public key material).  I'm also concerned about whether we should include a `+` in the label, but it appears to be a [legal character](https://github.com/RustCrypto/formats/blob/master/pem-rfc7468/src/grammar.rs#L28-L32).

alphan commented on sw/host/opentitanlib/src/crypto/spxplus.rs:
```diff
@@ -0,0 +1,164 @@
+
+    /// Verify a message signature, returning Ok(()) if the signature matches.
+    fn verify(&self, message: &[u8], sig: &SpxSignature) -> Result<()> {
+        spx::verify_detached_signature(
```
@jon-flatley FYI: _detached_ in this context means signatures that don't include the message. This is basically equivalent to (in form, of course) the RSA signature scheme with appendix that we are already using. @jadephilipoom please correct me if I'm wrong.

alphan reviewed (approved):
Thanks @jon-flatley ! We need to exclude `rsa_signature` as @cfrantz mentioned.

cfrantz commented on sw/host/opentitanlib/src/util/file.rs:
```diff
@@ -2,22 +2,74 @@
+        let len = file.read_to_end(&mut pem)?;
+
+        let mut decoder = Decoder::new(&pem[..len]).map_err(|e| anyhow!(e))?;
+        ensure!(
+            decoder.type_label() == Self::label(),
+            "PEM labels do not match \"{}\" != \"{}\"",
+            decoder.type_label(),
+            Self::label()
+        );
```
Can we create a formal error type for this condition?  You can use the `thiserror` crate (already in our deps):
```
use thiserror::Error;

#[derive(Debug, Error)]
enum MyError {
    #[error("PEM type error; expecting {0:?} but got {1:?}"]
    PemTypeError(String, String),
}

...
ensure!(a == b, MyError::PemTypeError(a, b));
```

cfrantz commented on sw/host/opentitanlib/src/util/file.rs:
```diff
@@ -2,22 +2,74 @@
+            Self::label()
+        );
+
+        let mut buf = vec![0u8; decoder.remaining_len()];
+        decoder
+            .decode(buf.as_mut_slice())
+            .map_err(|e| anyhow!(e))?;
```
If you enabled the `alloc` feature of the pem crate, you can reduce this to:
```
let mut buf = Vec::new();
decoder.decode_to_end(&mut buf);
```

jon-flatley replied:
I did consider this. If you think it's better to enable this feature in `pem_rfc7468` to drop this line I'm fine with that.

jon-flatley commented on sw/host/opentitanlib/src/util/file.rs:
```diff
@@ -2,22 +2,74 @@
+
+        let mut buf = [0u8; MAX_PEM_SIZE];
+        let mut encoder =
+            Encoder::new(Self::label(), LineEnding::LF, &mut buf).map_err(|e| anyhow!(e))?;
+        encoder.encode(&bytes).map_err(|e| anyhow!(e))?;
+        let len = encoder.finish().map_err(|e| anyhow!(e))?;
```
It seems to be necessary

```
error[E0277]: the trait bound `pem_rfc7468::Error: StdError` is not satisfied
  --> sw/host/opentitanlib/src/util/file.rs:38:35
   |
38 |         let len = encoder.finish()?;
   |                                   ^ the trait `StdError` is not implemented for `pem_rfc7468::Error`
```

jon-flatley replied:
Honestly I don't have the best understanding of error coercion. Could this be included as an additional error type in your `MyError` example below?

Something like this
```
#[derive(Debug, Error)]
enum PemError {
    #[error("PEM error: {0}")]
    InternalError(pem_rfc7468::Error),
    #[error("PEM type error; expecting {0} but got {1}")]
    LabelError(String, String),
}
```

And then
```
let len = encoder.finish().map_err(PemError::InternalError)?;
``` 

Still requires the `map_error()` call, so likely there is a better way.

cfrantz replied:
I didn't notice this earlier, but I think you need to enable the `std` feature on the PEM and RSA crates so that they'll `impl StdError` for their errors.

jon-flatley commented:
Thanks for the reviews, I think I've addressed everything at this point:

@alphan 
> There is no standard format yet. We floated binary blobs, hjson files (which can also encode type and other info) but PEM SGTM too.

Okay, the PEM stuff is included in a commit within this PR. It builds off of the binary format from before, so that part is unchanged. So the format is still an arbitrary bag of bytes, but now it's an arbitrary bag of bytes in PEM format.

> Yeah, we decided to exclude rsa_signature from the signed region.

Sounds good, this means the signed regions for both `rsa` and `spx` are the same correct? I've changed the functions in `Image` to reflect this.

@cfrantz 
I think I've addressed all of your comments, see my inline responses. I'm still not too sure the way I've adapted the error handing was what you had it mind, so please TAL at that.


I'm leaving #18041 as is until this piece gets merged.
